### PR TITLE
fix: Build stage2 stack and heap from suitable available memory.

### PR DESCRIPTION
**Pull request by bejeanmo**
Previously, Stage1B was building the Stage 2 stack and heap immediately below the FSP Reserved Memory resource descriptor region, assuming that this was a large available memory region. This assumption breaks when FSP resource descriptors move around, leading to unpredictable behavior.

This modification forces stage 1b to select a region for stack/heap from the FSP available memory descriptors that has adequete space for stack, heap, ACPI resources, and Payload reserved mem.

### CODE REVIEW

**Review by sindhuragrandhi — APPROVED** [review]

File: BootloaderCorePkg/Library/FspSupportLib/FspSupportLib.c [hunk 1]
```diff
@@ -33,6 +33,54 @@ GetFspNvsDataBuffer (
...
+  requested length.
+
+  @param  HobListPtr   A HOB list pointer.
+  @param  Length       A pointer to the requested length. If suitable
+                       descriptor is found, length will be updated with
+                       resource length.
+  @retval              Reserved region start address.  0 if this region does not exist.
+ */
+UINT64
+GetFspAvailableSystemMem(
```
Comment (rprangar) [inline BootloaderCorePkg/Library/FspSupportLib/FspSupportLib.c:47]: This function tries to find available memory of given length?

Then should it be renamed - as FindLoaderMemory or something like that without the FSP word in it?

File: BootloaderCorePkg/Library/FspSupportLib/FspSupportLib.c [hunk 2]
```diff
@@ -33,6 +33,54 @@ GetFspNvsDataBuffer (
...
+
+  /*
+   * Get the HOB list for processing
+   */
+  Hob.Raw = (VOID *)HobListPtr;
+
+  /*
+   * Collect memory ranges
+   */
+  while (!END_OF_HOB_LIST (Hob)) {
```
Comment (rprangar) [inline BootloaderCorePkg/Library/FspSupportLib/FspSupportLib.c:62]: Do we need to add any additional checks? Can this function find memory which is less than 1 MB?

File: BootloaderCorePkg/Library/FspSupportLib/FspSupportLib.c (hunk 1 above)
Comment (bejeanmo) [reply BootloaderCorePkg/Library/FspSupportLib/FspSupportLib.c:47]: It's looking in the FSP-M HOB list resource descriptors for available memory.

File: BootloaderCorePkg/Library/FspSupportLib/FspSupportLib.c (hunk 2 above)
Comment (bejeanmo) [reply BootloaderCorePkg/Library/FspSupportLib/FspSupportLib.c:62]: Guo and I are still discussing this fix. I think the three of us probably need to discuss this change some more.

**Comment by bejeanmo** [comment]
Updated implementation to use highest address region found for completeness, but this will be closed in favor of a temporary platform solution while we address this in FSP.

**Pull request closed without merging**